### PR TITLE
[C10D] Report detected failures when emitting collective end events.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109739
* #111239
* #111246

We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

### CODE REVIEW

File: test/distributed/test_hooks.py [hunk 1]
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
...
+        dhooks.register_collective_start_hook(coll_start)
+        dhooks.register_collective_end_hook(coll_end)
+
+        tensor = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        tensor_list = [torch.empty_like(tensor) for _ in range(self.world_size)]
+
+        dist.all_gather(tensor_list, tensor)
+
+        tensor2 = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        if self.rank == 2:
```
Comment (wconstab) [inline test/distributed/test_hooks.py:170]: are we assuming world_size == 2? is there a testing macro we can wrap to ensure that if so?

File: test/distributed/test_hooks.py [hunk 2]
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
...
+            cv.wait(20)
+
+        default_pg_name = dist.group.WORLD.group_name
+        self.assertEqual(2, len(starts))
+        self.assertEqual(2, len(ends))
+
+        def check_op(idx, coll_name, end_fail=False):
+            self.assertEqual(default_pg_name, starts[idx].pg_name)
+            self.assertEqual(self.backend_name, starts[idx].backend)
+            self.assertGreaterEqual(starts[idx].sequence_number, 0)
```
Comment (wconstab) [inline test/distributed/test_hooks.py:184]: why not assertEqual(sequence_number, idx)?

File: test/distributed/test_hooks.py [hunk 3]
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
...
+            self.assertGreaterEqual(starts[idx].timestamp, 0)
+            self.assertEqual(coll_name, starts[idx].operation)
+            self.assertEqual(starts[idx].error_message, None)
+            if end_fail:
+                self.assertIsNotNone(ends[idx].error_message, None)
```
Comment (wconstab) [inline test/distributed/test_hooks.py:189]: perhaps assert some known string is present in the error msg too?
(and that might also answer my question: what error are we expecting in this example?)

File: test/distributed/test_hooks.py (hunk 1 above)
Comment (kumpera) [reply test/distributed/test_hooks.py:171]: the test harness hardcode world size 4

File: test/distributed/test_hooks.py (hunk 3 above)
Comment (kumpera) [reply test/distributed/test_hooks.py:190]: good point, hopefully it will be reliable enough during testing.

**Comment by kumpera** [comment]
There's one major issue left on this PR: it doesn't work with gloo.

**Review by fduwjj — COMMENTED** [review]
How we ensure WatchDog does not abort the program and let us return error message?

File: torch/csrc/distributed/c10d/Backend.cpp [hunk 4]
```diff
@@ -18,6 +29,9 @@ void commonEventinit(
   evt.sequence_number = work.getSequencenumber();
   evt.operation = c10d::opTypeToString(work.retrieveOpType());
   evt.drop_count = 0;
+  // isCompleted is mutable :facepalm:
+  if (const_cast<Work&>(work).isCompleted() && !work.isSuccess())
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/Backend.cpp:33]: @fduwjj I think we should discuss this part

we're calling work.exception() which will query some cuda APIs.

we're doing this from the main watchdog thread.

i think that in today's watchdog design this is already something we're doing in the main loop, so i don't have a big reservation about it.

but if you're redesigning the loops to put cuda accesses on another thread, you'd also want to move this part.  would be good to think through how we'll handle the callback reporting error message safely, or else maybe not add the error to the callback in the first place.  (I think it should be possible to do though)

**Pull request closed without merging**